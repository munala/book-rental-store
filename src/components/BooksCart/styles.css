.cart-list-container {
  padding: 0px 20px;
  background-color: #eaeaea;
  border-radius: 20px;
  overflow: scroll;
  padding: 0px;
  width: 100%;
}

.cart-list {
  flex: 1;
  background-color: transparent;
  width: 100%;
}

.cart-list:last-child {
  border-bottom-width: 0px;
}

.cart-list-item {
  display: flex;
  background-color: transparent;
  border-style: none;
  border-bottom: 1px solid #ddd;
  padding: 20px 20px;
  justify-content: space-between;
}

.cart-list-item-select {
  background-color: #e1e1e1;
  padding: 5px 20px;
}

.cart-list-last-item {
  border-bottom-width: 0px;
}

.cart-button {
  width: 164px;
}

.cart-list-title {
  color: #7d6959;
  font-size: 20px;
  font-weight: bold;
  display: flex;
  flex-direction: row;
  align-items: flex-start;
}

.cart-list-item-right {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  flex-wrap: nowrap;
}

.cart-list-count {
  color: #555;
  font-size: 16px;
  font-weight: bold;
  display: flex;
  flex-direction: row;
  align-items: center;
  margin: 0px 20px;
}

.cart-list-count-value {
  width: 80px;
  margin-left: 10px;
}

.totals-row {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 20px 40px;
  background-color: #eaeaea;
  border-radius: 15px;
  overflow: hidden;
  padding: 10px 20px;
  margin: 20px;
}

.totals-row-left {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
}

.totals-row-title {
  color: #555;
  font-size: 20px;
  font-weight: bold;
}

.totals-row-value {
  color: #7d6959;
  font-size: 28px;
  font-weight: bold;
}

.hidden {
  opacity: 0;
}
