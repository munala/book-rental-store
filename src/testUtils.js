import React from "react";
import {
  createHistory,
  createMemorySource,
  LocationProvider
} from "@reach/router";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";

export const dispatchMock = jest.fn();

export const customer = {
  id: 1,
  email: "munalaoliverngoitsi@gmail.com"
};

export const loading = false;
export const error = null;

export const books = [
  {
    id: 1,
    title: "Jurassic Park",
    author: "Michael Crichton",
    genre: "action"
  }
];

export const message = "message";

export const LoadingContext = React.createContext(loading);
export const ErrorContext = React.createContext(error);
export const CustomerContext = React.createContext(customer);
export const BooksContext = React.createContext(books);

export const renderWithRouter = (
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) => {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history
  };
};

export const LoadingProvider = ({ children }) => (
  <LoadingContext.Provider value={{ customer, loadingDispatch: dispatchMock }}>
    {children}
  </LoadingContext.Provider>
);

export const ErrorProvider = ({ children }) => (
  <ErrorContext.Provider value={{ errorDispatch: dispatchMock }}>
    {children}
  </ErrorContext.Provider>
);

export const CustomerProvider = ({ children }) => (
  <CustomerContext.Provider value={{ customerDispatch: dispatchMock }}>
    {children}
  </CustomerContext.Provider>
);

export const BooksProvider = ({ children }) => (
  <BooksContext.Provider value={{ booksDispatch: dispatchMock }}>
    {children}
  </BooksContext.Provider>
);

export const App = ({ children }) => (
  <LoadingProvider>
    <ErrorProvider>
      <CustomerProvider>
        <BooksProvider>{children}</BooksProvider>
      </CustomerProvider>
    </ErrorProvider>
  </LoadingProvider>
);

export const MockCustomerComponent = ({
  customer: mockCustomer,
  logout,
  onSubmit,
  loginMode,
  message: mockMessage,
  setMessage
}) => {
  return (
    <div>
      <button
        id="onSubmit"
        onClick={() => {
          onSubmit(customer);
        }}
      />
      <button
        id="logout"
        onClick={() => {
          logout(customer);
        }}
      />
      <button
        id="message"
        onClick={() => {
          setMessage(message);
        }}
      />
      {mockCustomer && <div id="customerId">{mockCustomer.id}</div>}
      <div id="loginMode">{loginMode}</div>
      <div id="message">{mockMessage}</div>
    </div>
  );
};

export const MockBooksComponent = ({ books: mockBooks, loadBooks, search }) => {
  return (
    <div>
      <button
        id="loadBooks"
        onClick={() => {
          loadBooks();
        }}
      />

      <button
        id="search"
        onClick={() => {
          search();
        }}
      />

      {books?.map(book => {
        return (
          <div id={book.id} key={book.id}>
            {book.id}
          </div>
        );
      })}
    </div>
  );
};
