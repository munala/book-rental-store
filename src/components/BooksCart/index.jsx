import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import ListGroup from "react-bootstrap/ListGroup";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import {
  REGULAR_RENTAL,
  FICTION_RENTAL,
  NOVEL_RENTAL,
  REGULAR_MINIMUM_RENTAL,
  REGULAR_MINIMUM_DAYS,
  NOVEL_MINIMUM_RENTAL,
  NOVEL_MINIMUM_DAYS,
  FICTION_MINIMUM_RENTAL,
  FICTION_MINIMUM_DAYS,
  REGULAR_AFTER_MINIMUM_RENTAL,
  FICTION_AFTER_MINIMUM_RENTAL,
  NOVEL_AFTER_MINIMUM_RENTAL
} from "../../constants/rentals";
import "./styles.css";

const BooksCart = ({ cart, removeFromCart, clearCart, hide }) => {
  const [cartItems, setCartItems] = useState([]);
  const [total, setTotal] = useState(0);

  const rentalMappings = {
    regular: REGULAR_RENTAL,
    fiction: FICTION_RENTAL,
    novel: NOVEL_RENTAL
  };

  const rentalMinimumMappings = {
    regular: REGULAR_MINIMUM_RENTAL,
    fiction: FICTION_MINIMUM_RENTAL,
    novel: NOVEL_MINIMUM_RENTAL
  };

  const rentalAfterMinimumMappings = {
    regular: REGULAR_AFTER_MINIMUM_RENTAL,
    fiction: FICTION_AFTER_MINIMUM_RENTAL,
    novel: NOVEL_AFTER_MINIMUM_RENTAL
  };

  const rentalMinimumDaysMappings = {
    regular: REGULAR_MINIMUM_DAYS,
    fiction: FICTION_MINIMUM_DAYS,
    novel: NOVEL_MINIMUM_DAYS
  };

  // add quantity and number of days, default 1
  const processCartItems = useCallback(() => {
    const newCartItems = cart.map(item => {
      const existingItem = cartItems.find(cartItem => cartItem.id === item.id);

      if (existingItem) return existingItem;

      return {
        ...item,
        duration: 1,
        quantity: 1
      };
    });

    setCartItems(newCartItems);
  }, [cart]);

  const calculateRental = useCallback(
    item => {
      let totalCharge = 0;

      const numberOfDaysBeforeExtraCharge =
        rentalMinimumDaysMappings[item.genre];

      // normal charge before minimum number of days
      const chargesForMinimumDays =
        rentalMappings[item.genre] *
        item.quantity *
        numberOfDaysBeforeExtraCharge;

      // check if normal charge is greater than minimim otherwise use minimum
      totalCharge +=
        chargesForMinimumDays > rentalMinimumMappings[item.genre]
          ? chargesForMinimumDays
          : rentalMinimumMappings[item.genre];

      // check if duration has exceeded minimum than add additional charges for extra days
      if (item.duration > numberOfDaysBeforeExtraCharge) {
        const extraDays = item.duration - numberOfDaysBeforeExtraCharge;
        const extraCharge = extraDays * rentalAfterMinimumMappings[item.genre];

        totalCharge += extraCharge;
      }

      return totalCharge;
    },
    [cartItems]
  );

  useEffect(() => {
    processCartItems();
  }, [cart, processCartItems]);

  useEffect(() => {
    const totalAmount = cartItems.reduce(
      (previousTotal, item) => previousTotal + calculateRental(item),
      0
    );

    setTotal(totalAmount);
  }, [cartItems, setTotal]);

  const changeValue = ({ item, property, value }) => {
    const newCartItems = cartItems.map(cartItem => {
      return cartItem.id === item.id
        ? {
            ...cartItem,
            [property]: value < 0 ? 0 : value
          }
        : cartItem;
    });

    setCartItems(newCartItems);
  };

  return (
    <div>
      <div className="cart-list-container">
        <ListGroup as="ul" className="cart-list">
          <ListGroup.Item
            as="li"
            className={"cart-list-item cart-list-item-select"}
          >
            <div className="cart-list-title"></div>

            <div className="cart-list-item-right">
              <div className="cart-list-count">Books X</div>

              <div className="cart-list-count">
                <div className="cart-list-count-value">Days X</div>
              </div>

              <Button
                variant={"primary center-align cart-button"}
                onClick={hide}
              >
                Add More Books
              </Button>
            </div>
          </ListGroup.Item>

          {cartItems.map((item, index, list) => {
            return (
              <ListGroup.Item
                key={item.id}
                as="li"
                className={`cart-list-item ${
                  index === list.length - 1 ? "cart-list-last-item" : ""
                }`}
              >
                <div className="cart-list-title">{item.title}</div>
                <div className="cart-list-item-right">
                  <div className="cart-list-count">
                    <Form.Control
                      className="cart-list-count-value"
                      type="number"
                      aria-label=""
                      onChange={({ target: { value } }) =>
                        changeValue({ item, property: "quantity", value })
                      }
                      value={item.quantity}
                    />
                  </div>

                  <div className="cart-list-count">
                    <Form.Control
                      className="cart-list-count-value"
                      type="number"
                      aria-label=""
                      onChange={({ target: { value } }) =>
                        changeValue({ item, property: "duration", value })
                      }
                      value={item.duration}
                    />
                  </div>

                  <Button
                    variant={"outline-primary center-align cart-button"}
                    onClick={() => removeFromCart(item)}
                  >
                    Remove From Cart
                  </Button>
                </div>
              </ListGroup.Item>
            );
          })}
        </ListGroup>
      </div>

      <div className="totals-row">
        <div className="totals-row-left">
          <span className="totals-row-title">Total: </span>
          <span className="totals-row-value">${total}</span>
        </div>

        <Button
          variant={"primary center-align cart-button"}
          onClick={() => clearCart()}
        >
          Done
        </Button>
      </div>
    </div>
  );
};

BooksCart.propTypes = {
  removeFromCart: PropTypes.func.isRequired,
  clearCart: PropTypes.func.isRequired,
  hide: PropTypes.func,
  cart: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      author: PropTypes.string,
      genre: PropTypes.genre
    })
  ).isRequired
};

export default BooksCart;
