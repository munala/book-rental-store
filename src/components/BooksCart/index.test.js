import React from "react";
import renderer from "react-test-renderer";
import BooksCart from "./";
import { books } from "../../testUtils";

describe("BooksCart tests", () => {
  test("should render", () => {
    const removeFromCart = jest.fn();
    const clearCart = jest.fn();
    const hide = jest.fn();
    const cart = books;

    const tree = renderer
      .create(
        <BooksCart
          removeFromCart={removeFromCart}
          clearCart={clearCart}
          hide={hide}
          cart={cart}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
