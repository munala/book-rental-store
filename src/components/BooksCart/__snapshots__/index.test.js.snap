// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksCart tests should render 1`] = `
<div>
  <div
    className="cart-list-container"
  >
    <ul
      className="cart-list list-group"
      onKeyDown={[Function]}
    >
      <li
        className="cart-list-item cart-list-item-select list-group-item"
        data-rb-event-key={null}
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="cart-list-title"
        />
        <div
          className="cart-list-item-right"
        >
          <div
            className="cart-list-count"
          >
            Books X
          </div>
          <div
            className="cart-list-count"
          >
            <div
              className="cart-list-count-value"
            >
              Days X
            </div>
          </div>
          <button
            className="btn btn-primary center-align cart-button"
            disabled={false}
            onClick={[MockFunction]}
            type="button"
          >
            Add More Books
          </button>
        </div>
      </li>
    </ul>
  </div>
  <div
    className="totals-row"
  >
    <div
      className="totals-row-left"
    >
      <span
        className="totals-row-title"
      >
        Total: 
      </span>
      <span
        className="totals-row-value"
      >
        $
        0
      </span>
    </div>
    <button
      className="btn btn-primary center-align cart-button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Done
    </button>
  </div>
</div>
`;
